frame/name-service/src/lib.rs

```diff
@@ -0,0 +1,487 @@
+		_,
+		Blake2_128Concat,
+		NameHash,
+		Registration<T::AccountId, BalanceOf<T>, T::BlockNumber>,
+	>;
+
+	/// This resolver maps name hashes to an account
+	#[pallet::storage]
+	#[pallet::getter(fn resolve)]
+	pub(super) type Resolvers<T: Config> = StorageMap<_, Blake2_128Concat, NameHash, T::AccountId>;
+
+	// Your Pallet's events.
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		// TODO: Potentially make events more lightweight
+		/// A new `Commitment` has taken place.
+		Committed { sender: T::AccountId, who: T::AccountId, hash: CommitmentHash },
+		/// A new `Registration` has taken added.
+		NameRegistered { name_hash: NameHash, owner: T::AccountId, expiry: T::BlockNumber },
+		/// A `Registration` has been transferred to a new owner.
+		Transfer { from: T::AccountId, to: T::AccountId },
+		/// A `Registration` has been renewed.
+		NameRenewed { name_hash: NameHash, expires: T::BlockNumber },
+		/// An address has been set for a name hash to resolve to.
+		AddressSet { name_hash: NameHash, address: T::AccountId },
+		/// An address was deregistered.
+		AddressDeregistered { name_hash: NameHash },
+	}
+
+	// Your Pallet's error messages.
+	#[pallet::error]
+	pub enum Error<T> {
+		/// This commitment hash already exists in storage.
+		AlreadyCommitted,
+		/// It has not passed the minimum waiting period to reveal a commitement.
+		TooEarlyToReveal,
+		/// The commitment cannot yet be removed. Has not expired.
+		CommitmentNotExpired,
+		/// This commitment does not exist.
+		CommitmentNotFound,
+		/// This registration does not exist.
+		RegistrationNotFound,
+		/// The sender is not the registration owner.
+		NotRegistrationOwner,
+		/// This registration has expired.
+		RegistrationExpired,
+		/// This registration has not yet expired.
+		RegistrationNotExpired,
+	}
+
+	// Your Pallet's callable functions.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		// TODO: Should we allow registration on behalf of?
+		#[pallet::weight(0)]
+		pub fn commit(
+			origin: OriginFor<T>,
+			who: T::AccountId,
+			commitment_hash: CommitmentHash,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+
+			ensure!(!Commitments::<T>::contains_key(commitment_hash), Error::<T>::AlreadyCommitted);
+
+			let block_number = frame_system::Pallet::<T>::block_number();
+			let deposit = T::CommitmentDeposit::get();
+
+			T::Currency::reserve(&sender, deposit)?;
+
+			let commitment = Commitment { who: who.clone(), when: block_number, deposit };
+
+			Commitments::<T>::insert(commitment_hash, commitment);
+			Self::deposit_event(Event::<T>::Committed { sender, who, hash: commitment_hash });
+
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn reveal(
+			origin: OriginFor<T>,
+			name: Vec<u8>,
+			secret: u64,
+			periods: u32,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			let commitment_hash = sp_io::hashing::blake2_256(&(&name, secret).encode());
```

Comment:
We definitely need a way to canonicalize names.  
ENS uses [UTS46](https://unicode.org/reports/tr46/) which is standard for URLs.

Reply:
this happens on the front-end, not in the pallet.

Reply:
Not expert on unicode but is it possible for A and A' both canonicalize into A'? Then you have duplicated registration.

Reply:
Yes that is the problem. The backend needs to canonicalize and provide one source of truth for all frontends.  
Assume that `Alice.eth` and `alice.eth` are different addresses or not?  
The backend decides.

Reply:
ENS is known broken. https://discuss.ens.domains/t/zero-width-characters-pose-a-security-risk-and-existential-threat-to-ens/476/3

Reply:
It seems that ENS still handles this problem in the front end still. We can add some unicode checking inside the pallet, but seems very "unblockchain".

Reply:
My point is ENS is broken. Please don't use it as a working reference.
I am not sure what's the right solution here because unicode itself is also a it broken.
I guess one other solution is require all supported frontend to validate the url and instead of canonicalize it, just reject them all the alternative form. But this will be a footgun at some stage as some client will failed to follow it.

Reply:
Another way is to don't do validation by default (to avoid extra computation) but expose a validation call that can be triggered by anyone and an invalid name will be slashed and deposit sent to the challenger.

Reply:
I wouldn't say ENS is broken as it does what it intended to do; allow anyone to register a domain using any character set Unicode supports.

To play devils advocate, limiting such functionality could also limit innovation and potential use cases. Such as registering a secondary name (ross.dot with a ZWC) as a private namehash you'd only use for a couple of contacts or something.

What may be a good strategy front-end is to force users to type in an address rather than copy and paste it, before saving to an address book. From there they can copy and paste into a search field, with tap to copy on the resulting addresses.
Emojis are CTRL + CMD + Space on a Mac, Windows + . on Windows. 
Users also can dive into settings and change keyboard input for their preferred character set.
Trivially easy on mobile.

For reference, ZWs [are used](https://unicode-explorer.com/c/200B):
>... to indicate word boundaries to text processing systems when using scripts that do not use explicit spacing, or after characters (such as the slash) that are not followed by a visible space but after which there may nevertheless be a line break. It is also used with languages without visible space between words, for example...

And emojis of course. So it seems something this multi-purpose cannot be squeezed out of the service when falling short of re-writing how Unicode works / an alternative - and even then front ends would need to translate some ad-hoc format.

Definitely not a trivial problem if it were to be solved pallet-side.